localization/graph_localizer/src/feature_point_graph_values.cc

```diff
@@ -0,0 +1,121 @@
+                                                            const localization_common::Time timestamp) const {
+  return boost::none;
+}
+
+boost::optional<lc::Time> FeaturePointGraphValues::OldestTimestamp() const { return boost::none; }
+
+boost::optional<lc::Time> FeaturePointGraphValues::LatestTimestamp() const { return boost::none; }
+
+boost::optional<lc::Time> FeaturePointGraphValues::SlideWindowNewOldestTime() const { return boost::none; }
+
+bool FeaturePointGraphValues::HasFeature(const lm::FeatureId id) const { return (feature_id_key_map_.count(id) > 0); }
+
+boost::optional<gtsam::Key> FeaturePointGraphValues::FeatureKey(const lm::FeatureId id) const {
+  if (!HasFeature(id)) return boost::none;
+  return feature_id_key_map_.at(id);
+}
+
+gtsam::Key FeaturePointGraphValues::CreateFeatureKey() const { return sym::F(++feature_key_index_); }
+
+gtsam::KeyVector FeaturePointGraphValues::FeatureKeys() const {
+  gtsam::KeyVector feature_keys;
+  for (const auto& feature_id_key_pair : feature_id_key_map_) {
+    feature_keys.emplace_back(feature_id_key_pair.second);
+  }
+  return feature_keys;
+}
+
+bool FeaturePointGraphValues::AddFeature(const lm::FeatureId id, const gtsam::Point3& feature_point,
+                                         const gtsam::Key& key) {
+  if (HasFeature(id)) {
+    LogError("AddFeature: Feature already exists.");
+    return false;
+  }
+
+  if (values().exists(key)) {
+    LogError("AddFeature: Key already exists in values.");
+  }
+
+  feature_id_key_map_.emplace(id, key);
+  values().insert(key, feature_point);
+  return true;
+}
+
+int FeaturePointGraphValues::NumFeatures() const { return feature_id_key_map_.size(); }
+
+gtsam::KeyVector FeaturePointGraphValues::OldKeys(const localization_common::Time oldest_allowed_time,
+                                                  const gtsam::NonlinearFactorGraph& factors) const {
+  using ProjectionFactor = gtsam::GenericProjectionFactor<gtsam::Pose3, gtsam::Point3>;
+  gtsam::KeyVector old_features;
+  for (const auto& feature_id_key_pair : feature_id_key_map_) {
+    const auto& key = feature_id_key_pair.second;
+    int num_factors = 0;
+    for (const auto& factor : factors) {
+      // Only consider projection factors for min num feature factors
+      const auto projection_factor = dynamic_cast<const ProjectionFactor*>(factor.get());
+      if (!projection_factor) continue;
+      if (factor->find(key) != factor->end()) {
+        ++num_factors;
+        // if (num_factors >= params_.min_num_factors_per_feature) break;
+      }
+    }
+
+    // if (num_factors < params_.min_num_factors_per_feature) {
+    if (num_factors <= 0) {
+      old_features.emplace_back(key);
+    }
+  }
+  return old_features;
+}
+
+void FeaturePointGraphValues::RemoveOldFeatures(const gtsam::KeyVector& old_keys) {
+  for (const auto& key : old_keys) {
+    // TODO(rsoussan): test this
```

Comment:
Did we test this?

Reply:
This was deprioritized because its for sliding window bundle adjustment, which is pretty experimental right now